cpp/src/arrow/array/builder_dict.cc

```diff
@@ -40,6 +40,16 @@ using internal::checked_cast;
 // ----------------------------------------------------------------------
 // DictionaryType unification
 
+template <typename T, typename Out = void>
```

Comment:
@pitrou I ran into brittleness with the way the template specialization was working in this file. I refactored things to more explicitly use `std::enable_if`